### ISSUE TITLE
Build agent in AKS is faling to deploy

### ISSUE THREAD

**Issue by vivuu1989** [microsoft/azure-pipelines-agent#3999]
We are using AzureDevops Server(2020) and trying to install ADO buildagents in AKS service(1.23.8) Which is having Istio (1.13.5
)enabled.

We were refering the Microsoft Document and below are the options tried.

First used Docker file

```
FROM ubuntu:18.04

# To make it easier for build and release pipelines to run apt-get,
# configure apt to not require confirmation (assume the -y argument by default)
ENV DEBIAN_FRONTEND=noninteractive
RUN echo "APT::Get::Assume-Yes \"true\";" > /etc/apt/apt.conf.d/90assumeyes

RUN apt-get install -y --no-install-recommends \
    ca-certificates \
    curl \
    jq \
    git \
    iputils-ping \
    libcurl4 \
    libicu60 \
    libunwind8 \
    netcat \
    libssl1.0 \
  && rm -rf /var/lib/apt/lists/*

RUN curl -LsS https://aka.ms/InstallAzureCLIDeb | bash \
  && rm -rf /var/lib/apt/lists/*

# Can be 'linux-x64', 'linux-arm64', 'linux-arm', 'rhel.6-x64'.
ENV TARGETARCH=linux-x64

WORKDIR /azp

COPY ./start.sh .
RUN chmod +x start.sh

ENTRYPOINT ["./start.sh"]
```

Deployment File

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: azdevops-deployment
  labels:
    app: azdevops-agent
spec:
  replicas: 1 #here is the configuration for the actual agent always running
  selector:
    matchLabels:
      app: azdevops-agent
  template:
    metadata:
      labels:
        app: azdevops-agent
    spec:
      containers:
      - name: kubepodcreation
        image: qrdevcontainers.azurecr.io/vsts/dockeragent:1.0
        env:
          - name: AZP_URL
            valueFrom:
              secretKeyRef:
                name: azdevops
                key: AZP_URL
          - name: AZP_TOKEN
            valueFrom:
              secretKeyRef:
                name: azdevops
                key: AZP_TOKEN
          - name: AZP_POOL
            valueFrom:
              secretKeyRef:
                name: azdevops
                key: AZP_POOL
        volumeMounts:
        - mountPath: /var/run/docker.sock
          name: docker-volume
      volumes:
      - name: docker-volume
        hostPath:
          path: /var/run/docker.sock
```
		  
		  
		  
		  
Getting Below error

```
1. Determining matching Azure Pipelines agent...
2. Downloading and extracting Azure Pipelines agent...
curl: (35) OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to vstsagentpackage.azureedge.net:443

gzip: stdin: unexpected end of file
tar: Child returned status 1
tar: Error is not recoverable: exiting now
```

**Comment by Roman-Shchukin** [comment]
It looks like a network problem. Could you check by pinging this address or something like that?

**Comment by vivuu1989** [comment]
Tried telnet to the given address on port 443 from the aks node and it's
getting connected.

**Comment by Roman-Shchukin** [comment]
You did it from inside docker container?

**Comment by vivuu1989** [comment]
It's running on aks cni based network concept. So telnet is tried from the
node itself as containers and nodes in same network in azure cni network
model.

**Comment by vivuu1989** [comment]
when I tried to run the above built image as a docker conatiner agent from one of the buildagent machine itself, getting below error

docker run -e AZP_URL=<Azure DevOps instance> -e AZP_TOKEN=<PAT token> -e AZP_AGENT_NAME=mydockeragent dockeragent:latest



```
>> Connect:

Connecting to server ...

>> Register Agent:

Scanning for tool capabilities.
Connecting to the server.
Successfully replaced the agent
Testing agent connection.
2022-10-25 12:54:30Z: Settings Saved.

4. Running Azure Pipelines agent...
chmod: cannot access './run-docker.sh': No such file or directory
Cleanup. Removing Azure Pipelines agent...
Removing agent from the server
Error reported in diagnostic logs. Please examine the log for more details.
    - /azp/_diag/Agent_20221025-125431-utc.log
Failed: Removing agent from the server
TF400813: Resource not available for anonymous access. Client authentication required.
```

Here I am using same MS provided start.sh and Dockerfile to build the image.  Why its failing when its running as a standalone container itself.  What file exactly its looking for? (chmod: cannot access './run-docker.sh': No such file or directory)

When I used the same image in the replicaset in AKS cluster  getting different error, where as both the AKS container subnet and the Machine subnet are same and have conectivity to ADO server and "vstsagentpackage.azureedge.net"

```
1. Determining matching Azure Pipelines agent...
2. Downloading and extracting Azure Pipelines agent...
curl: (35) OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to vstsagentpackage.azureedge.net:443

gzip: stdin: unexpected end of file
tar: Child returned status 1
tar: Error is not recoverable: exiting now
```

**Comment by sergey-koryshev** [comment]
@vivuu1989 could you please provide  link to documentation you followed?

**Comment by vivuu1989** [comment]
Here is the document.

https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops#linux

**Comment by vivuu1989** [comment]
@sergey-koryshev , Yes, you are right, I was referring some wrong package and that caused the issue. But After that I could successfully create the deployment with the package 2.170.1 vsts agent but faced below issue.

We were able to successfully add the deployment to Azuredevops Agent pool and could execute the pipeline on them by following the [Microsoft docs][1]. But Its throwing strange error related to dynatrace for all the tasks which is getting executed inside them..

    /opt/dynatrace/oneagent/agent/bin/1.251.xxxxxxxxxx/any/nodejs/nodejsagent.js:2611
      const e = Buffer.allocUnsafe(8), d = 2 ** 32;
                                              ^
    
    SyntaxError: Unexpected token *
        at createScript (vm.js:56:10)
        at Object.runInThisContext (vm.js:97:10)
        at Module._compile (module.js:549:28)
        at Object.Module._extensions..js (module.js:586:10)
        at Module.load (module.js:494:32)
        at tryModuleLoad (module.js:453:12)
        at Function.Module._load (module.js:445:3)
        at Module.require (module.js:504:17)
        at module.js:676:12
        at Array.forEach (native)
    ##[error]Exit code 1 returned from process: file name '/azp/externals/node/bin/node', arguments '"/azp/_work/_tasks/PublishBuildArtifacts_2ff763a7-ce83-4e1f-bc89-0ae63477cebe/1.158.3/publishbuildartifacts.js"'.


  [1]: https://learn.microsoft.com/en-us/azure/devops/pipelines/agents/docker?view=azure-devops#linux


    Note:- Same image is tried with the Docker standalone container and where the pipelines are getting executed successfully. Where as when we trigger the pipelines to the AKS based agents (using the same docker image), getting the above syntax error. 

Adding the logs which we got from Build-Agent
********************************************

/azp/_diag/Worker_20221108-092937-utc.log





    [2022-11-08 09:29:40Z INFO AgentPluginManager] Loaded task plugin id 'ecxxxxxxx-xxxx-xxxx-xxx-x-xxxxxxe' (main).
    [2022-11-08 09:29:40Z INFO WorkerCommandManager] Enable plugin internal command extension.
    [2022-11-08 09:29:40Z INFO HostContext] Well known directory 'Bin': '/azp/bin'
    [2022-11-08 09:29:40Z INFO HostContext] Well known directory 'Root': '/azp'
    [2022-11-08 09:29:40Z INFO HostContext] Well known directory 'Work': '/azp/_work'
    [2022-11-08 09:29:40Z INFO HostContext] Well known directory 'Bin': '/azp/bin'
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper] Starting process:
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper]   File name: '/azp/bin/Agent.PluginHost'
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper]   Arguments: 'task "Agent.Plugins.Repository.CheckoutTask, Agent.Plugins"'
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper]   Working directory: '/azp/_work'
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper]   Require exit code zero: 'True'
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper]   Encoding web name: utf-8 ; code page: '65001'
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper]   Force kill process on cancellation: 'False'
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper]   Redirected STDIN: 'True'
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper]   Persist current code page: 'False'
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper]   Keep redirected STDIN open: 'False'
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper]   High priority process: 'False'
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper] Failed to update oom_score_adj for PID: 2515.
    [2022-11-08 09:29:40Z INFO ProcessInvokerWrapper] System.UnauthorizedAccessException: Access to the path '/proc/2515/oom_score_adj' is denied.
    ---> System.IO.IOException: Permission denied
       --- End of inner exception stack trace ---
       at System.IO.FileStream.WriteNative(ReadOnlySpan`1 source)
       at System.IO.FileStream.FlushWriteBuffer()
       at System.IO.FileStream.Dispose(Boolean disposing)
       at System.IO.Stream.Close()
       at System.IO.StreamWriter.CloseStreamFromDispose(Boolean disposing)
       at System.IO.StreamWriter.Dispose(Boolean disposing)
       at System.IO.TextWriter.Dispose()
       at System.IO.File.WriteAllText(String path, String contents)
       at Microsoft.VisualStudio.Services.Agent.Util.ProcessInvoker.WriteProcessOomScoreAdj(Int32 processId, Int32 oomScoreAdj)

I again tried with , vsts 2.211.1 version and after that none of the pipelines are getting scheduled in the vsts agents eventhough the agents are running and listening for the jobs.

**Comment by sergey-koryshev** [comment]
@vivuu1989, sorry for late answer. Do you still experience the issue? 

Issue with `PublishBuildArtifacts` occurs due to the fact that the task was run on wrong version of `Node` by some reasons. It requires `10` or `16` version but logs says it was picked up `Node` of version `6`. To further investigation we need to see full logs of agent and example of YAML of affected pipeline. Can you please attach them?

As for error `Failed to update oom_score_adj for PID: 2515`, I believe this is related to old version of agent you use, we have seen such issues before and it was fixed in version `2.196.1` but you mentioned you used version of agent `2.170.1`. So I just suggest you to use last version to rid the issue.

**Comment by sergey-koryshev** [comment]
I close the issue because of inactivity. Please feel free to reactivate it or create new one.
